### PR TITLE
End to end tests

### PR DESCRIPTION
**Pull request by KarthikBhaskara**
## Description
Added end to end tests for the SDK.

The e2e tests are required to run serially so I've added a new script for it -
```npm run test:e2e```.

## Reviewer Checklist

- [ ]  All endpoints have been tested
- [ ] All existing and new tests run with the ```npm test``` command without throwing any errors.

### CODE REVIEW

File: test/end-to-end/helpers/index.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    // add username/password and submit
+    .setValue('input[id=username]', email)
+    .setValue('input[id=password]', 'password')
+    .click('button[id=approval-button]')
+    .url((currentUrl) => {
+      if (currentUrl.value.startsWith('http://localhost:4040/callback')) {
+        // if we have skipped the permissions, we can extract the code
+        test(getCodeFromUrl(currentUrl.value));
+        browser
+          .pause(5000)
```
Comment (RogWilco) [inline test/end-to-end/helpers/index.js]: Just curious, what does this pause (and the similar one below) do?

File: package.json [hunk 2]
```diff
@@ -48,7 +51,10 @@
     "cross-env": "^5.1.3",
     "eslint": "^4.18.1",
     "eslint-config-smartcar": "^2.0.0",
+    "express": "^4.16.3",
```
Comment (RogWilco) [inline package.json]: Is this dependency still needed?

File: package.json [hunk 3]
```diff
@@ -48,7 +51,10 @@
...
+    "incito": "^1.1.1",
     "jsdoc-to-markdown": "^3.0.0",
+    "nightwatch": "^0.9.21",
```
Comment (RogWilco) [inline package.json]: Shouldn't this be a dev-dependency?

File: package.json [hunk 4]
```diff
@@ -48,7 +51,10 @@
     "cross-env": "^5.1.3",
     "eslint": "^4.18.1",
     "eslint-config-smartcar": "^2.0.0",
+    "express": "^4.16.3",
+    "incito": "^1.1.1",
```
Comment (RogWilco) [inline package.json]: Same as above, is this dependency still needed?

File: test/end-to-end/helpers/index.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
+'use strict';
+
+const Chance = require('chance');
+
+const chance = new Chance();
+
+const helpers = {};
+
+helpers.getAuthClientParams = function() {
```
Comment (RogWilco) [inline test/end-to-end/helpers/index.js]: Any reason this is a function instead of just an object literal?

File: test/end-to-end/helpers/index.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    .waitForElementVisible('div[class=content]', 1000)
+    .click('a[href^="https://mock.smartcar.com"]')
+    // login dialog is rendered
+    .waitForElementVisible('input[id=username]', 1500)
+    // add username/password and submit
+    .setValue('input[id=username]', email)
+    .setValue('input[id=password]', 'password')
+    .click('button[id=approval-button]')
+    .url((currentUrl) => {
+      if (currentUrl.value.startsWith('http://localhost:4040/callback')) {
```
Comment (RogWilco) [inline test/end-to-end/helpers/index.js]: Since you already have this string defined earlier in this file, why not reference it here instead of duplicating the string?

File: test/end-to-end/helpers/index.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+helpers.getAuthClientParams = function() {
+  return {
+    clientId: 'e922556a-7d4f-4168-88cd-059276044798',
+    clientSecret: '79c07401-d3b2-48c0-8407-dc19c4ece7ff',
+    redirectUri: 'http://localhost:4040/callback',
+    development: true,
+  };
+};
+
+const getCodeFromUrl = function(uri) {
```
Comment (RogWilco) [inline test/end-to-end/helpers/index.js]: Nit: Naming mismatch. Either name the function `getCodeFromUri` or name the parameter variable `url`.

File: test/end-to-end/helpers/index.js (hunk 1 above)
Comment (KarthikBhaskara) [reply test/end-to-end/helpers/index.js]: We pass in the test as a callback to nightwatch but it looks like Nightwatch does not run it asynchronously, or rather, does not wait for the execution to complete. Until we find a way to do that, we will simply pause giving enough time for the callback to complete execution.

File: test/end-to-end/helpers/index.js (hunk 5 above)
Comment (KarthikBhaskara) [reply test/end-to-end/helpers/index.js]: No reason, my logic was if we want to change the params based on what test is calling it, it gives us that extensibility. Thoughts?

File: test/end-to-end/helpers/index.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+const Chance = require('chance');
+const querystring = require('querystring');
+
+const chance = new Chance();
+
+const helpers = {};
+
+helpers.getAuthClientParams = function() {
+  return {
+    clientId: 'e922556a-7d4f-4168-88cd-059276044798',
```
Comment (GregChan) [inline test/end-to-end/helpers/index.js]: It was my bad for pushing these up, but you should make these environment variables. Ask @RogWilco about which client id and secret to use. I've rolled the secret for this so no one can use it.

File: package.json [hunk 9]
```diff
@@ -25,19 +25,21 @@
...
-    "cover": "nyc npm test -s",
+    "cover": "nyc npm run test:unit -s",
     "cover:integration": "nyc npm test:integration -s",
     "release": "semantic-release",
     "jsdoc": "jsdoc -c .jsdoc.json .",
     "docs": "mkdir -p doc && jsdoc2md --example-lang js --template doc/.template.hbs --files .docs.js index.js lib/* | sed 's/[ \t]*$//' > doc/readme.md"
   },
   "dependencies": {
     "bluebird": "^3.5.1",
+    "convict": "^4.3.2",
```
Comment (gurpreetatwal) [inline package.json]: this should be a dev dep

File: test/.eslintrc.js [hunk 10]
```diff
@@ -4,4 +4,11 @@ module.exports = {
   parserOptions: {
     ecmaVersion: 2017,
   },
+  rules: {
+    'space-before-function-paren': ['error', {
+      anonymous: 'never',
+      named: 'never',
+      asyncArrow: 'always',
+    }],
```
Comment (gurpreetatwal) [inline test/.eslintrc.js]: why was this changed?

File: test/config/index.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
+'use strict';
+
+/* eslint strict: ['error', 'global'] */
+/* global require, __dirname, module */
```
Comment (gurpreetatwal) [inline test/config/index.js]: I don't think these are needed here

File: test/end-to-end/auth-client.js [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+  context.client = nightwatch.initClient(config.get('nightwatch'));
+  context.browser = context.client.api();
+});
+
+test('exchangeCode', (t) => {
+  const client = new smartcar.AuthClient(getAuthClientParams());
+
+  const authUrl = client.getAuthUrl();
+
+  return runAuthFlow(context.client, context.browser, authUrl)
```
Comment (gurpreetatwal) [inline test/end-to-end/auth-client.js]: nit: you could just make the test an async function, mixing the two is a bit weird imo...

File: test/end-to-end/auth-client.js [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+        const keys = Object.keys(access).sort();
+        const hasKeys = _.isEmpty(
+          _.xor(keys, [
+            'accessToken',
+            'expiration',
+            'refreshExpiration',
+            'refreshToken',
+          ])
+        );
+        t.true(hasKeys);
```
Comment (gurpreetatwal) [inline test/end-to-end/auth-client.js]: @snggeng and I played around with this for awhile and found that doing `t.deepEqual(_.xor(_.keys(<object>), [<expected-keys>]), []);`  produces the best diff when they don't match

File: test/end-to-end/auth-client.js [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+        t.pass();
```
Comment (gurpreetatwal) [inline test/end-to-end/auth-client.js]: don't need

File: test/end-to-end/auth-client.js [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+      } catch (err) {
+        t.fail();
```
Comment (gurpreetatwal) [inline test/end-to-end/auth-client.js]: you should just let the error bubble up, this makes the test harder to debug, no?

File: test/end-to-end/auth-client.js [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
```
Comment (gurpreetatwal) [inline test/end-to-end/auth-client.js]: same comments on this test case as above

File: test/end-to-end/helpers/index.js [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+  };
+};
+
+const getCodeFromUri = function(uri) {
+  // this helper functions assumes the structure of the uri
+  // is valid and contains the required code
+  const firstIndexOfParams = uri.indexOf('?') + 1;
+  const params = uri.substring(firstIndexOfParams);
+
+  return querystring.parse(params).code;
```
Comment (gurpreetatwal) [inline test/end-to-end/helpers/index.js]: we could use Node's built in URL parsers: https://nodejs.org/docs/latest-v6.x/api/url.html#url_class_urlsearchparams
that we don't have to make any assumptions

File: test/end-to-end/helpers/index.js [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+      .waitForElementVisible('div[class=permissions]', 5000)
+      .click('button[id=approval-button]')
+      .url((currentUrl) => {
+        // extract code from callback url
+        code = getCodeFromUri(currentUrl.value);
+      })
+      .end();
+
+    client.start(() => {
+      resolve(code);
```
Comment (gurpreetatwal) [inline test/end-to-end/helpers/index.js]: nit: don't need the brackets

File: test/end-to-end/index.js [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+test.before(() => {
+  context.client = nightwatch.initClient(config.get('nightwatch'));
+  context.browser = context.client.api();
+
+  const client = new smartcar.AuthClient(getAuthClientParams());
+
+  const authUrl = client.getAuthUrl();
+
+  return runAuthFlow(context.client, context.browser, authUrl)
+    .then(async (code) => {
```
Comment (gurpreetatwal) [inline test/end-to-end/index.js]: nit: don't mix then chains and async

File: test/end-to-end/index.js [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+      context.accessToken = access.accessToken;
+    });
+});
+
+test('getVehicleIds', async (t) => {
+  t.notThrows(async () => await smartcar.getVehicleIds(context.accessToken));
+});
+
+test('getUserId', async (t) => {
+  t.notThrows(async () => await smartcar.getUserId(context.accessToken));
```
Comment (gurpreetatwal) [inline test/end-to-end/index.js]: should we be doing some other assertions other than just `notThrows`? Feels really coarse

File: test/end-to-end/vehicle.js [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+test.before(() => {
+  context.client = nightwatch.initClient(config.get('nightwatch'));
+  context.browser = context.client.api();
+
+  const client = new smartcar.AuthClient(getAuthClientParams());
+
+  const authUrl = client.getAuthUrl();
+
+  return runAuthFlow(context.client, context.browser, authUrl)
+    .then(async (code) => {
```
Comment (gurpreetatwal) [inline test/end-to-end/vehicle.js]: same nit about then chains and promises

File: test/end-to-end/vehicle.js [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+test('vehicle unlock', async (t) => {
+  t.notThrows(async () => await context.vehicle.unlock());
+});
+
+test('vehicle permissions', async (t) => {
+  t.notThrows(async () => await context.vehicle.permissions());
+});
+
+test('vehicle disconnect', async (t) => {
+  t.notThrows(async () => await context.vehicle.disconnect());
```
Comment (gurpreetatwal) [inline test/end-to-end/vehicle.js]: same question here about assertions

File: test/end-to-end/index.js (hunk 20 above)
Comment (KarthikBhaskara) [reply test/end-to-end/index.js]: Resolved irl

File: test/end-to-end/vehicle.js (hunk 22 above)
Comment (KarthikBhaskara) [reply test/end-to-end/vehicle.js]: Resolved irl

File: test/end-to-end/vehicle.js [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+test('vehicle unlock', async(t) => {
+  t.notThrows(async() => await context.vehicle.unlock());
+});
+
+test('vehicle permissions', async(t) => {
+  t.notThrows(async() => await context.vehicle.permissions());
+});
+
+test('vehicle disconnect', async(t) => {
+  t.notThrows(async() => await context.vehicle.disconnect());
```
Comment (gurpreetatwal) [inline test/end-to-end/vehicle.js]: I don't think these tests work correctly as written reason being that the `t.notThrows` is not being `await`'d 

The vehicles methods return promises and we await that in the callback function and then `t.notThrows` awaits the callback function, but nothing waits for the promise that `t.notThrows` returns.

Since `t.notThrows` accepts a promise, the callback isn't really needed either, as such you should do one of the following:

```js
test('name', async(t) => {
  await t.notThrows(context.vehicle.method());
});
```

or

```js
test('name', (t) => {
  return t.notThrows(context.vehicle.method());
});
```

There are two tests in index.js which do something similar please fix those as well.
